collect/src/main/kotlin/com/certora/collect/TreapListNode.kt

```diff
@@ -0,0 +1,428 @@
+        goRight = { this.with(right = right!!.removeNodeAt(it)) },
+    )
+
+    private fun getNodeAt(index: Int): TreapListNode<E> = traverseToIndex(
+        index,
+        found = { this },
+        goLeft = { left!!.getNodeAt(it) },
+        goRight = { right!!.getNodeAt(it) },
+    )
+
+    private fun firstNode(): TreapListNode<E> = left?.firstNode() ?: this
+    private fun lastNode(): TreapListNode<E> = right?.lastNode() ?: this
+
+    private fun leftIndex(parentIndex: Int) = parentIndex - 1 - (right?.size ?: 0)
+    private fun rightIndex(parentIndex: Int) = parentIndex + 1 + (left?.size ?: 0)
+    private fun rootIndex() = leftIndex(size)
+
+    private fun indexOfFirstNode(thisIndex: Int, element: E): Int? =
+        left?.indexOfFirstNode(left.leftIndex(thisIndex), element)
+        ?: thisIndex.takeIf { this.elem == element }
+        ?: right?.indexOfFirstNode(right.rightIndex(thisIndex), element)
+
+    private fun indexOfLastNode(thisIndex: Int, element: E): Int? =
+        right?.indexOfLastNode(right.rightIndex(thisIndex), element)
+        ?: thisIndex.takeIf { this.elem == element }
+        ?: left?.indexOfLastNode(left.leftIndex(thisIndex), element)
+
+    private fun setNodeAt(index: Int, element: E): TreapListNode<E> = traverseToIndex(
+        index,
+        found = { this.with(elem = element) },
+        goLeft = { this.with(left = left!!.setNodeAt(it, element)) },
+        goRight = { this.with(right = right!!.setNodeAt(it, element)) },
+    )
+
+    private fun removeAllNodes(predicate: (E) -> Boolean): TreapListNode<E>? = when {
+        predicate(this.elem) -> left?.removeAllNodes(predicate) append right?.removeAllNodes(predicate)
+        else -> this.with(left = left?.removeAllNodes(predicate), right = right?.removeAllNodes(predicate))
+    }
+
+    private fun updateNodes(transform: (E) -> E?): TreapListNode<E>? {
+        val newLeft = left?.updateNodes(transform)
+        val newElem = transform(elem)
+        val newRight = right?.updateNodes(transform)
+        return when {
+            newElem == null -> newLeft append newRight
+            else -> this.with(left = newLeft, right = newRight, elem = newElem)
+        }
+    }
+
+    private fun updateNodesIndexed(thisIndex: Int, transform: (Int, E) -> E?): TreapListNode<E>? {
+        val newLeft = left?.updateNodesIndexed(left.leftIndex(thisIndex), transform)
+        val newElem = transform(thisIndex, elem)
+        val newRight = right?.updateNodesIndexed(right.rightIndex(thisIndex), transform)
+        return when {
+            newElem == null -> newLeft append newRight
+            else -> this.with(left = newLeft, right = newRight, elem = newElem)
+        }
+    }
+
+    private fun forEachNode(action: (E) -> Unit) {
+        left?.forEachNode(action)
+        action(elem)
+        right?.forEachNode(action)
+    }
+
+    private fun forEachNodeIndexed(thisIndex: Int, action: (Int, E) -> Unit) {
```

Comment:
I think this can be removed, and the public function written as:
```
fun forEachElementIndexed(action : (Int, E) -> Unit) {
   var i = 0
   forEachNode {
      action(i++, it)
   }
}
```

Also, I believe this would also be much clearer and just as good as the current implementation:
```
override fun indexOf(element: E): Int? {
    forEachElementIndexed { i, e -> 
        if (element == e) return i
    }
    return null
}
```
for the reverse case, maybe it's best to just implement reverse traversal? it's much simpler code-wise than these calculations. At least to me....

Reply:
Like elsewhere, the recursive descent here avoids allocating a closure.  It probably doesn't make a significant difference for large lists, but for small ones this way should be noticeably faster.